Runtime implementation section (#60)
golemfactory/yagna-docs: requestor-tutorials/service-development/service-example-2-managed-erigon.md

### Erigon runtime

First, lets discuss what is the purpose of the Erigon runtime to gain some high level overview.

* accepts the Ethereum network's name from the requestor agent,
* starts Erigon binaries to sync with the chosen network and serve Erigon RPC endpoint,
* generates credentials for basic authentication which is implemented with nginx
* credentials are returned to the requestor agent so user of the service can connect to the Erigon RPC.

Please refer to the "Provider's node" in the above Architecture overview diagram.

### Implementation

You might want to review simpler [example runtime](https://github.com/golemfactory/ya-runtime-sdk/blob/main/examples/example-runtime/src/main.rs) available in the [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk/) repository.

Erigon runtime declares two structures

* `ErigonConf` - containing runtime configuration with default values,
* `ErigonRuntime` - derives from `RuntimeDef` implements the required methods from `ya-runtime-sdk::Runtime` trait

> At least this methods from the `Runtime` trail has to be implemented but SDK provides default implementation to several others (TODO: link runtime SDK docs)

Reviewer: ```suggestion
At least these methods from the `Runtime` trail have to be implemented but SDK provides default implementation to several others (TODO: link runtime SDK docs)
```

> * `deploy` - usually called before any command is send to the runtime, but it's possible to call it explicitly by agent SDKs.

Reviewer: This sentence can be misunderstood. `deploy` and `start` are mandatory and always called in sequence

   It's run only once per activity, does not accept parameters (but there are plans to add them in the future release), return value is not passed to the requestor agent. In the Erigon runtime this method creates directories for Erigon data.

> * `start` - called once per activity, accepts parameters, return value is not passed to the requestor agent. Analogically with `deploy` if not run explicitly via the SDK is run without parameters before any command is send to the runtime. In the Erigon runtime we use `start` method to receive the `network` parameter from the requestor, start Erigon's binaries and generate new credentials for nginx basic auth.

Reviewer: > called once per activity

Technically it can be called multiple times per activity, currently SDKs do not do that.

> return value is not passed to the requestor agent

What kind of return value do you have in mind? The output **should** be visible to the requestor (if not, please open a new issue)

* `run_command` - called every time the corresponding agent SDK [`run`](runhttps://github.com/golemfactory/yagna-docs/blob/master/yapapi/api-reference.md#run) is called on the `WorkContext` object. Result can be obtained from the `WorkContext`'s results array after sequence of commands is executed by `commit`. In the Erigon runtime this method is used to provide Erigon service's credentials to the agent.

* `stop` - can be called explicitly by the requestor agent or automatically due to agreement termination. In the Erigon runtime `stop` method is used to kill Erigon processes started in `start`.

Note that the difference between `deploy` and `start` regarding the Erigon runtime might not be obvious. It's easier to think of them in terms of VM runtimes where `deploy` is needed for runtime preparation, e.g. download the image, after `start` runtime should be ready to run the commands.

### Extentions

Above methods corresponds to [`CLI::Command` enum](https://github.com/golemfactory/ya-runtime-sdk/blob/main/ya-runtime-sdk/src/cli.rs#L11). One can extend the default `CLI` by 

* derive `StructOps` on the `CLI` struct
* decorate the runtime struct with `#[cli(<struct_name>)]`

> * custom CLI arguments will be passed to each command by `Runtime::Context`

Reviewer: ```suggestion
* custom CLI arguments will be passed to each command by `runtime::Context`
```

`runtime` is a module

## Runtime implementation

Here we demonstrate how self-contained Exe-Unit runtimes can be implemented, tested and finally plugged into the Provider's daemon.

:warning: Please be aware that custom Exe-Unit runtime doesn't provide the level of isolation comparable with gvmi image. It's just executable running with the same privileges that daemon does.

### Erigon runtime

First, lets discuss what is the purpose of the Erigon runtime to gain some high level overview.

* accepts the Ethereum network's name from the requestor agent,
* starts Erigon binaries to sync with the chosen network and serve Erigon RPC endpoint,
* generates credentials for basic authentication which is implemented with nginx
* credentials are returned to the requestor agent so user of the service can connect to the Erigon RPC.

Please refer to the "Provider's node" in the above Architecture overview diagram.

> ### Implementation

Reviewer 2: This section needs snippets of actual code with description. See how e.g. [simple-service](https://handbook.golem.network/requestor-tutorials/service-development/service-example-1-simple-service) is written.

* `start` - called once per activity, accepts parameters, return value is not passed to the requestor agent. Analogically with `deploy` if not run explicitly via the SDK is run without parameters before any command is send to the runtime. In the Erigon runtime we use `start` method to receive the `network` parameter from the requestor, start Erigon's binaries and generate new credentials for nginx basic auth.

* `run_command` - called every time the corresponding agent SDK [`run`](runhttps://github.com/golemfactory/yagna-docs/blob/master/yapapi/api-reference.md#run) is called on the `WorkContext` object. Result can be obtained from the `WorkContext`'s results array after sequence of commands is executed by `commit`. In the Erigon runtime this method is used to provide Erigon service's credentials to the agent.

* `stop` - can be called explicitly by the requestor agent or automatically due to agreement termination. In the Erigon runtime `stop` method is used to kill Erigon processes started in `start`.

> Note that the difference between `deploy` and `start` regarding the Erigon runtime might not be obvious. It's easier to think of them in terms of VM runtimes where `deploy` is needed for runtime preparation, e.g. download the image, after `start` runtime should be ready to run the commands.

Reviewer 2: ```suggestion
{% hint style="info" %}
Note that the difference between `deploy` and `start` regarding the Erigon runtime might not be obvious. It's easier to think of them in terms of VM runtimes where `deploy` is needed for runtime preparation, e.g. download the image, after `start` runtime should be ready to run the commands.
{% endhint %}
```

Another option is `RuntimeMode::Command` where each command is a separate invocation of the runtime binary. 

### Runtime configuration

During runtime startup the configuration file located in `$HOME/.local/share/ya-runtime-erigon/ya-runtime-erigon.json` is parsed and is initialising the instance of `ErigonConf` structure. If the file does not exists it will be created and initialized with default values during the first start of the runtime.

#### Example configuration file

```json
{
  "public_addr": "https://0.erigon.golem.network:8545",
  "data_dir": "/data/erigon",
  "passwd_tool_path": "htpasswd",
  "passwd_file_path": "/etc/nginx/erigon_htpasswd",
  "password_default_length": 15,
  "erigon_http_addr": "127.0.0.1",
  "erigon_http_port": "8555"
}
```

### Testing the runtime

Once needed SDK methods are implemented we can give our runtime a try in the [debugger](https://github.com/golemfactory/ya-runtime-dbg/).
For example our Erigon runtime can be executed in the debugger shell running the following command.

```bash
ya-ya-runtime-dbg \
  --runtime ./target/debug/ya-runtime-erigon \
  --workdir ./target/debug \
  --start-arg '{"network": "goerli"}'
```

> Please mind it requires Erigon binaries present in the same directory (all needed binaries for Ubuntu Linux can be downloaded from the [binary release](https://github.com/golemfactory/yagna-service-erigon/releases/tag/ya-runtime-erigon-v0.1.0)). Also `htpasswd` from the `apache-tools` is needed to be present in the `PATH` but for the local testing you can mock it with `echo` changing configuration file property `"passwd_tool_path": "echo",`

Reviewer 2: ```suggestion
{% hint style="info" %}
Please mind it requires Erigon binaries present in the same directory (all needed binaries for Ubuntu Linux can be downloaded from the [binary release](https://github.com/golemfactory/yagna-service-erigon/releases/tag/ya-runtime-erigon-v0.1.0)). Also `htpasswd` from the `apache-tools` is needed to be present in the `PATH` but for the local testing you can mock it with `echo` changing configuration file property `"passwd_tool_path": "echo",`
{% endhint %}
```

{% endhint %}

### Erigon runtime

First, let's discuss what is the purpose of the Erigon runtime to gain some high level overview. The runtime:

* Accepts an Ethereum network's name (e.g. 'mainnet' or 'rinkeby') from the requestor agent,
* Starts Erigon binaries to sync with the chosen network and serve Erigon RPC endpoint,
* Generates credentials for basic authentication which is implemented with nginx,
* Returns the credentials to the requestor agent so the user of the service can connect to the Erigon RPC.

Please refer to the "Provider's node" in the above Architecture overview diagram.

### Implementation

{% hint style="info" %}
Before diving into this section you might want to review simpler [example runtime](https://github.com/golemfactory/ya-runtime-sdk/blob/main/examples/example-runtime/src/main.rs) available in the [ya-runtime-sdk](https://github.com/golemfactory/ya-runtime-sdk/) repository.
{% endhint %}

Erigon runtime declares two structures

* `ErigonConf` - containing runtime configuration with default values,
* `ErigonRuntime` - derives from `RuntimeDef` implements the required methods from `ya-runtime-sdk::Runtime` trait

> At least these methods from the `Runtime` trait have to be implemented but SDK provides default implementation to several others (TODO: link runtime SDK docs)

Reviewer 2: `TODO: link runtime SDK docs`
Link still missing.

> * `deploy` - It's run only once per activity, does not accept parameters (but there are plans to add them in the future release), return value is not passed to the requestor agent.

Reviewer 2: For clarity, I strongly recommend making `deploy`, `start`, `run_command` and `stop` into sub-sections of 'Implementation'. These single-bullet lists don't stand out enough.

  
In the Erigon runtime we use `start` method to receive the `network` parameter from the requestor, start Erigon's binaries and generate new credentials for nginx basic auth.
As above, we'll review the code more deeply.

First, we receive the input from the caller from context `ctx.cli.command.args()` provided by SDK's `CommandCLI` trait `Command` enum.
Arguments to start are passed as a `&Vec<String>`, we're interested only of the first of them.
`get_chain_from_config` is retrieving the network value from the json and converts it to the `Network` variant, to assert it's within supported networks.
We're also setting its lowercased String value in the `erigon_chain` property for the later use. 
```rust
fn start<'a>(&mut self, ctx: &mut Context<Self>) -> OutputResponse<'a> {
    let chain = get_chain_from_config(ctx.cli.command.args().into_iter().next());
    let chain = chain.to_string().to_lowercase();
    self.erigon_chain = Some(chain.clone());
```

Next, we're generating new password for the nginx's basic auth. We're generating random string,

> set it with the `passwd_tool` and remember it in the `erigon_password` service's property.

Reviewer 2: ```suggestion
set it with the `passwd_tool` and store it in the `erigon_password` service's property.
```

```rust
// Generate user & password entry with passwd tool
    let rng = thread_rng();
    let password: String = rng.sample_iter(Alphanumeric)
        // ...
        .collect();

    let _ = generate_password_file(
        &ctx.conf.passwd_tool_path,
        &ctx.conf.passwd_file_path,
        &password,
    )
    .expect("Unable to create passwd file");
    self.erigon_password = Some(password);
```

> What's left is to spawn `erigon` and `rpcdaemon` processes and remember handles to them so we can kill them when the service will be stopped.

Reviewer 2: ```suggestion
What's left is to spawn `erigon` and `rpcdaemon` processes and store handles to them so we can kill them when the service will be stopped.
```

            .arg("--chain")
            .arg(&chain),
        &data_dir,
        &[""; 0],
    )
    .expect("Erigon: Failed to spawn");

    let rpcd_pid = spawn_process(...);

    self.erigon_pid = Some(erigon_pid);
    self.rpcdaemon_pid = Some(rpcd_pid);
```

The `start`'s return value is serving only troubleshooting purposes as the Erigon agent ignores it but it can be reviewed in the debugger session.

{% hint style="info" %}
Note that the difference between `deploy` and `start` regarding the Erigon runtime might not be obvious. It's easier to think of them in terms of VM runtimes where `deploy` is needed for runtime preparation, e.g. download the image, after `start` runtime should be ready to run the commands.
{% endhint %}

* `run_command` - called every time the corresponding agent SDK [`run`](runhttps://github.com/golemfactory/yagna-docs/blob/master/yapapi/api-reference.md#run) is called on the `WorkContext` object. 
  Result can be obtained from the `WorkContext`'s results array after sequence of commands is executed by `commit`. 
  
In the Erigon runtime this method is used to provide Erigon service's credentials to the agent.

> Here we're just forming the json string response and pass it as a command output, see the call of `run_ctx.stdout(...)` below.

Reviewer 2: ```suggestion
Here we're just forming the JSON string response and passing it as a command output, see the call of `run_ctx.stdout(...)` below.
```

## Runtime implementation

Here we demonstrate how self-contained Exe-Unit runtimes can be implemented, tested and finally plugged into the Provider's daemon.

{% hint style="warning" %}

> Please be aware that a custom runtime doesn't provide the level of isolation comparable with a virtual machine. It's just an executable running with the same privileges as yagna daemon.

Reviewer 3: ```suggestion
Please be aware that a custom runtime doesn't provide the level of isolation comparable with a virtual machine. It's just an executable running with the same privileges as the yagna daemon.
```

{% endhint %}

### Erigon runtime

> First, let's discuss what is the purpose of the Erigon runtime to gain some high level overview. The runtime:

Reviewer 3: ```suggestion
First, let's discuss what the purpose of the Erigon runtime is to gain some high level overview. The runtime:
```
